crates/phactory/src/prpc_service.rs

```diff
@@ -1807,6 +1839,213 @@ impl<Platform: pal::Platform + Serialize + DeserializeOwned> PhactoryApi for Rpc
+        let (ecdh_pubkey, encrypted_key) = key_share::encrypt_secret_to(
+            &my_identity_key,
+            &[b"worker_key_handover"],
+            &ecdh_pubkey.0,
+            &my_identity_key.dump_secret_key(),
+            &iv,
+        )
+        .map_err(from_debug)?;
+        let encrypted_key = EncryptedKey {
+            ecdh_pubkey: sr25519::Public(ecdh_pubkey),
+            encrypted_key,
+            iv,
+        };
+        let runtime_state = phactory.runtime_state()?;
+        let genesis_block_hash = runtime_state.genesis_block_hash;
+        let encrypted_worker_key = EncryptedWorkerKey {
+            genesis_block_hash,
+            para_id: runtime_state.para_id,
+            dev_mode,
+            encrypted_key,
+        };
+
+        let worker_key_hash = blake2_256(&encrypted_worker_key.encode());
+        let mut report_data = [0u8; 64];
+        report_data[..32].copy_from_slice(&worker_key_hash);
+        let remote_target_info = sgx_api_lite::target_info_from_report(recv_local_report);
+        let report = sgx_api_lite::report(&remote_target_info, &report_data).map_err(|_| {
+            from_display("Failed to create attestation report for worker key handover")
+        })?;
+        let la_report = sgx_api_lite::encode(&report).to_vec();
+        Ok(pb::DcapHandoverWorkerKey::new(
+            encrypted_worker_key,
+            la_report,
+        ))
+    }
+
+    // Dcap WorkerKey Handover Client
+    async fn dcap_handover_accept_challenge(
+        &mut self,
+        request: pb::DcapHandoverChallenge,
+    ) -> RpcResult<pb::DcapHandoverChallengeResponse> {
+        let mut phactory = self.lock_phactory(false, true)?;
+
+        // generate and save tmp key only for key handover encryption
+        let handover_key = crate::new_sr25519_key();
+        let handover_ecdh_key = handover_key
+            .derive_ecdh_key()
+            .expect("should never fail with valid key; qed.");
+        let ecdh_pubkey = phala_types::EcdhPublicKey(handover_ecdh_key.public());
+        phactory.handover_ecdh_key = Some(handover_ecdh_key);
+
+        let challenge = request.decode_challenge().map_err(from_display)?;
+        let challenge_handler = DcapChallengeHandlerInfo {
+            challenge: challenge.clone(),
+            ecdh_pubkey,
+        };
+        let handler_hash = blake2_256(&challenge_handler.encode());
+        let mut report_data = [0u8; 64];
+        report_data[..32].copy_from_slice(&handler_hash);
+        // generate local attestation report to ensure the handover pRuntimes are on the same machine
+        let la_report = {
+            let its_target_info = unsafe {
+                sgx_api_lite::decode(&challenge.sgx_target_info)
+                    .map_err(|_| from_display("Invalid client sgx target info"))?
+            };
+            // the report data does not matter since we only care about the origin
+            let report = sgx_api_lite::report(its_target_info, &report_data)
```

Comment:
Does this mean an IAS pRuntime can also get key from DCAP pRuntime?

Reply:
No, IAS pruntime validates RA to each other during handover, while dcap pruntime server won't generate RA report. So, an IAS pruntime won't accept an handover from DCAP pruntime.